### PR TITLE
fastfetch: init at 1.12.2

### LINKED ISSUE #176799: Add fastfetch
**Project description**
fastfetch is a [neofetch] like tool for fetching system information and displaying them in a pretty way. It is written in c to achieve much better performance,

**Metadata**

* source URL: https://github.com/LinusDierheimer/fastfetch
* license: mit, bsd, gpl2+ ,:  mit
* platforms: unix, linux, darwin, : linux

### PR DESCRIPTION
**Pull request by Gerg-L**
###### Description of changes
Added fastfetch, essentially a faster version of neofetch.
Closes #176799. #176819

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Gerg-L** [comment]
Made this PR more like the previous one
Added an update script and a test

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -1599,6 +1599,10 @@ with pkgs;
 
   etlegacy = callPackage ../games/etlegacy { lua = lua5_4; };
 
+  fastfetch = callPackage ../tools/misc/fastfetch { };
```
Comment (khaneliman) [inline pkgs/top-level/all-packages.nix:1602]:
```suggestion
  fastfetch = darwin.apple_sdk_11_0.callPackage ../tools/misc/fastfetch {
    inherit (darwin.apple_sdk_11_0.frameworks) AppKit Cocoa CoreDisplay CoreVideo CoreWLAN DisplayServices
                                               Foundation IOBluetooth MediaRemote OpenCL;
  };
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    ++ lib.optional (withFullDeps || enableSqlite) sqlite
+    ++ lib.optional (withFullDeps || enableVulkan) vulkan-loader
+    ++ lib.optional (withFullDeps || enableWayland) wayland
+    ++ lib.optional (withFullDeps || enableX11) libxcb
+    ++ lib.optional (withFullDeps || enableXfce) xfce.xfconf;
+
+  buildInputs =
+    finalAttrs.runtimeDependencies
+    ++ lib.optional (withFullDeps || enableOpenCL) opencl-headers
+    ++ lib.optional (withFullDeps || enableX11) libXrandr;
```
Comment (khaneliman) [inline pkgs/tools/misc/fastfetch/default.nix:101]:
```suggestion
    ++ lib.optional (withFullDeps || enableX11) libXrandr
      ++ lib.optionals stdenv.isDarwin [
    AppKit
    Cocoa
    CoreDisplay
    CoreVideo
    CoreWLAN
    DisplayServices
    Foundation
    IOBluetooth
    MediaRemote
    OpenCL
    darwin.moltenvk
  ];
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  #nativeBuildInputs
+  cmake,
+  makeWrapper,
+  pkg-config,
+  #buildInputs
+  dbus,
+  dconf,
+  glib,
+  pciutils,
+  zlib,
```
Comment (khaneliman) [inline pkgs/tools/misc/fastfetch/default.nix:16]:
```suggestion
  zlib,
  #darwinInputs
  , AppKit
, Cocoa
, CoreDisplay
, CoreVideo
, CoreWLAN
, DisplayServices
, Foundation
, IOBluetooth
, MediaRemote
, OpenCL
, darwin
```

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -1599,6 +1599,10 @@ with pkgs;
...
+
+  fastfetch-full = callPackage ../tools/misc/fastfetch { withFullDeps = true; };
```
Comment (khaneliman) [inline pkgs/top-level/all-packages.nix:1604]:
```suggestion
  fastfetch-full = darwin.apple_sdk_11_0.callPackage ../tools/misc/fastfetch { 
  withFullDeps = true;
   inherit (darwin.apple_sdk_11_0.frameworks) AppKit Cocoa CoreDisplay CoreVideo CoreWLAN DisplayServices
                                               Foundation IOBluetooth MediaRemote OpenCL;
    };
```

**Review by khaneliman — CHANGES_REQUESTED** [review]
Changes to support darwin

File: pkgs/tools/misc/fastfetch/default.nix [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  cmakeFlags = [
+    "-DCMAKE_INSTALL_SYSCONFDIR=${placeholder "out"}/etc"
+  ];
+
+  passthru = {
+    updateScript = nix-update-script {};
+    tests.version = testers.testVersion {
+      package = finalAttrs.finalPackage;
+      command = "fastfetch -v";
+      version = "fastfetch -v ${finalAttrs.version} (${stdenv.hostPlatform.linuxArch})";
```
Comment (khaneliman) [inline pkgs/tools/misc/fastfetch/default.nix:112]: any way to make this test linux specific and add another for darwin ?

File: pkgs/tools/misc/fastfetch/default.nix [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    [
+      dbus
+      dconf
+      glib
+      pciutils
+      zlib
+    ]
+    ++ lib.optional (withFullDeps || enableChafa) chafa
+    ++ lib.optional (withFullDeps || enableImageMagick) imagemagick_light
+    ++ lib.optional (withFullDeps || enableNM) networkmanager
```
Comment (khaneliman) [inline pkgs/tools/misc/fastfetch/default.nix:87]: most of these are linux specific and should probably add a flag for linux dependencies

File: pkgs/tools/misc/fastfetch/default.nix [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    wrapProgram $out/bin/flashfetch \
+      --prefix LD_LIBRARY_PATH : "${finalAttrs.ldLibraryPath}"
+  '';
+
+  meta = with lib; {
+    description = "Like neofetch, but much faster because written in C";
+    inherit (finalAttrs.src.meta) homepage;
+    license = licenses.mit;
+    platforms = platforms.linux;
+    maintainers = with maintainers; [gerg-l];
```
Comment (khaneliman) [inline pkgs/tools/misc/fastfetch/default.nix:130]:
```suggestion
    maintainers = with maintainers; [gerg-l khaneliman];
```

**Comment by khaneliman** [comment]
@Gerg-L mind if i just add a commit specific for supporting darwin to this?

**Comment by Gerg-L** [comment]
@khaneliman we can co-auther the commit, I'm available to work on this together in a few hours

**Comment by khaneliman** [comment]
I created 2 commits to cherry-pick https://github.com/khaneliman/nixpkgs/commits/fastfetch. One of the dependencies for the aarch64-darwin requires a rebase of the PR branch.

**Review by khaneliman — APPROVED** [review]

**Review by khaneliman — CHANGES_REQUESTED** [review]
need to rebase on master and nixpkgs-fmt

**Comment by Gerg-L** [comment]
> need to rebase on master and nixpkgs-fmt 

Need to fix the `callPackage`'s

**Comment by khaneliman** [comment]
nixpkgs-review successful
- aarch64-darwin
- x86_64-darwin
- x86_64-linux

> > need to rebase on master and nixpkgs-fmt
> 
> Need to fix the `callPackage`'s

Yeah it's missing a dependency that is on current master branch and not on yours

**Comment by Gerg-L** [comment]
The issue was a `withAllDeps` as well

**Review by khaneliman — APPROVED** [review]

File: pkgs/tools/misc/fastfetch/default.nix [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+, OpenCL
+,
+}:
+stdenv.mkDerivation (finalAttrs: {
+  pname = "fastfetch";
+  version = "1.12.2";
+
+  src = fetchFromGitHub {
+    owner = "fastfetch-cli";
+    repo = finalAttrs.pname;
```
Comment (khaneliman) [inline pkgs/tools/misc/fastfetch/default.nix:51]: apparently we want to use a string literal here instead of the finalAttrs.pname

**Review by khaneliman — CHANGES_REQUESTED** [review]
minor tweak for repo = pname

**Review by khaneliman — APPROVED** [review]

**Review by khaneliman — APPROVED** [review]

**Comment by khaneliman** [comment]
@drupol  @AndersonTorres any chance either of you could take a look at this tomorrow sometime?

**Comment by AndersonTorres** [comment]
Do you wanna roast?

**Comment by khaneliman** [comment]
> Do you wanna roast?

Not sure what you mean lol

File: pkgs/tools/misc/fastfetch/default.nix [hunk 9]
```diff
@@ -0,0 +1,128 @@
+{ fetchFromGitHub
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:1]: The order is
-  lib
- stdenv
- fetchers
- all the rest in ASCIIbetical order
- if there are dependencies on Darwin, they will be the last ones

File: pkgs/tools/misc/fastfetch/default.nix [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+, lib
+, nix-update-script
+, stdenv
+, testers
+, #nativeBuildInputs
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:6]: This is not a tutorial on how to write your first  expression. Remove these useless comments.

File: pkgs/tools/misc/fastfetch/default.nix [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  };
+
+  nativeBuildInputs = [
+    cmake
+    makeWrapper
+    pkg-config
+  ];
+
+  buildInputs =
+    [
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:63]:
```suggestion
  buildInputs = [
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+      chafa
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:64]: Indent this accordingly.

File: pkgs/tools/misc/fastfetch/default.nix [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+      xfce.xfconf
+      zlib
+    ]
+    ++ lib.optionals stdenv.isDarwin [
+      AppKit
+      Cocoa
+      CoreDisplay
+      CoreVideo
+      CoreWLAN
+      darwin.moltenvk
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:92]:
```suggestion
      moltenvk
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    wrapProgram $out/bin/flashfetch \
+      --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath finalAttrs.buildInputs}"
+  '';
+
+  meta = {
+    description = "Like neofetch, but much faster because written in C";
+    inherit (finalAttrs.src.meta) homepage;
+    license = lib.licenses.mit;
+    platforms = lib.platforms.all;
+    maintainers = with lib.maintainers; [ gerg-l khaneliman ];
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:125]:
```suggestion
    maintainers = with lib.maintainers; [ gerg-l khaneliman ];
    platforms = lib.platforms.unix;
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    mainProgram = "fastfetch";
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:126]:
```suggestion
```

It is the default already.

File: pkgs/top-level/all-packages.nix [hunk 16]
```diff
@@ -1678,6 +1678,11 @@ with pkgs;
 
   etlegacy = callPackage ../games/etlegacy { lua = lua5_4; };
 
+  fastfetch = darwin.apple_sdk_11_0.callPackage ../tools/misc/fastfetch {
+    inherit (darwin.apple_sdk_11_0.frameworks) AppKit Cocoa CoreDisplay CoreVideo CoreWLAN DisplayServices
+      Foundation IOBluetooth MediaRemote OpenCL;
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:1683]:
```suggestion
    inherit (darwin.apple_sdk_11_0.frameworks)
       AppKit Cocoa CoreDisplay CoreVideo CoreWLAN DisplayServices
      Foundation IOBluetooth MediaRemote OpenCL;
```

I prefer one per line, but this should be sufficient

File: pkgs/tools/misc/fastfetch/default.nix [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+      IOBluetooth
+      MediaRemote
+      OpenCL
+    ];
+
+  cmakeFlags = [
+    "-DCMAKE_INSTALL_SYSCONFDIR=${placeholder "out"}/etc"
+  ];
+
+  passthru = {
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:104]: `passthru` goes right before `meta`.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/fastfetch/default.nix (hunk 9 above)
Comment (khaneliman) [reply pkgs/tools/misc/fastfetch/default.nix:1]: Where is the documentation for some of this stuff? I couldn't seem to find it from the CONTRIBUTING.md or https://nixos.org/manual/nixpkgs/stable/#reviewing-contributions-new-packages

**Comment by khaneliman** [comment]
@Gerg-L https://github.com/khaneliman/nixpkgs/commit/c60f3c829275b434f57125edd649ab069a632a97

File: pkgs/tools/misc/fastfetch/default.nix (hunk 15 above)
Comment (Gerg-L) [reply pkgs/tools/misc/fastfetch/default.nix:126]: leaving it in case #246386 merges

File: pkgs/top-level/all-packages.nix [hunk 18]
```diff
@@ -1678,6 +1678,12 @@ with pkgs;
 
   etlegacy = callPackage ../games/etlegacy { lua = lua5_4; };
 
+  fastfetch = darwin.apple_sdk_11_0.callPackage ../tools/misc/fastfetch {
+    inherit (darwin.apple_sdk_11_0.frameworks)
+    AppKit Cocoa CoreDisplay CoreVideo CoreWLAN DisplayServices
+    Foundation IOBluetooth MediaRemote moltenvk OpenCL;
+  };
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:1685]:
```suggestion
    inherit (darwin) moltenvk;
  };
```

**Comment by Gerg-L** [comment]
alright double and triple checked

File: pkgs/tools/misc/fastfetch/default.nix (hunk 15 above)
Comment (AndersonTorres) [reply pkgs/tools/misc/fastfetch/default.nix:121]: Ok then.

File: pkgs/tools/misc/fastfetch/default.nix [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+, vulkan-loader
+, wayland
+, xfce
+, zlib
+, AppKit
+, Cocoa
+, CoreDisplay
+, CoreVideo
+, CoreWLAN
+, moltenvk
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:33]:
```suggestion
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+, DisplayServices
+, Foundation
+, IOBluetooth
+, MediaRemote
+, OpenCL
+,
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:39]:
```suggestion
, moltenvk
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+}:
+stdenv.mkDerivation (finalAttrs: {
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:41]:
```suggestion
}:

stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+    xfce.xfconf
+    zlib
+  ]
+  ++ lib.optionals stdenv.isDarwin [
+    AppKit
+    Cocoa
+    CoreDisplay
+    CoreVideo
+    CoreWLAN
+    moltenvk
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:87]:
```suggestion
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+    DisplayServices
+    Foundation
+    IOBluetooth
+    MediaRemote
+    OpenCL
+  ];
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:93]:
```suggestion
    moltenvk
  ];
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by khaneliman — APPROVED** [review]

File: pkgs/tools/misc/fastfetch/default.nix [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+, CoreDisplay
+, CoreVideo
+, CoreWLAN
+, DisplayServices
+, Foundation
+, IOBluetooth
+, MediaRemote
+, moltenvk
+, OpenCL
+,
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:39]:
```suggestion
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+, Cocoa
+, CoreDisplay
+, CoreVideo
+, CoreWLAN
+, DisplayServices
+, Foundation
+, IOBluetooth
+, MediaRemote
+, moltenvk
+, OpenCL
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:38]:
```suggestion
, OpenCL
, moltenvk
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+    Cocoa
+    CoreDisplay
+    CoreVideo
+    CoreWLAN
+    DisplayServices
+    Foundation
+    IOBluetooth
+    MediaRemote
+    moltenvk
+    OpenCL
```
Comment (AndersonTorres) [inline pkgs/tools/misc/fastfetch/default.nix:93]:
```suggestion
    OpenCL
    moltenvk
```

ASCIIbetical order. Uppercase goes first.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by khaneliman — APPROVED** [review]

File: pkgs/tools/misc/fastfetch/default.nix [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+, cmake
+, dbus
+, dconf
+, glib
+, imagemagick_light
+, libglvnd
+, libpulseaudio
+, libxcb
+, libXrandr
+, makeWrapper
```
Comment (drupol) [inline pkgs/tools/misc/fastfetch/default.nix:14]:
```suggestion
, makeBinaryWrapper
```

File: pkgs/tools/misc/fastfetch/default.nix [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+  src = fetchFromGitHub {
+    owner = "fastfetch-cli";
+    repo = "fastfetch";
+    rev = finalAttrs.version;
+    hash = "sha256-l9fIm7+dBsOqGoFUYtpYESAjDy3496rDTUDQjbNU4U0=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    makeWrapper
```
Comment (drupol) [inline pkgs/tools/misc/fastfetch/default.nix:54]:
```suggestion
    makeBinaryWrapper
```

**Review by drupol — CHANGES_REQUESTED** [review]
Hi!

A minor change and it's good to go!

**Review by drupol — APPROVED** [review]
Thanks!

**Review by AndersonTorres — APPROVED** [review]

**Comment by Gerg-L** [comment]
Thank you drupol!

**Comment by hervyqa** [comment]
Great. Thank you!!

**Comment by Izorkin** [comment]
@Gerg-L is it possible to make a variant without GUI components to be able to install on remote servers?

**Pull request merged**